### PR TITLE
Replication unit tests

### PR DESCRIPTION
**Pull request by Nao-ris**

### CODE REVIEW

**Review by MatthieuClochardDoctolib — CHANGES_REQUESTED** [review]
General thoughts:

- should we move unsafe to the function signature instead of opening unsafe block for public api (not sure what is the standard in Rust)
- maybe we should check the visibility of our fn, not sure everything exposed as pub should be

Other remarks outside the scope of this PR:
- https://github.com/Nao-ris/couchbase-lite-rust/blob/e03d508163710ef8c68513b33496a0ec5e99ef96/src/replicator.rs#L49 and https://github.com/Nao-ris/couchbase-lite-rust/blob/e03d508163710ef8c68513b33496a0ec5e99ef96/src/replicator.rs#L56 and https://github.com/Nao-ris/couchbase-lite-rust/blob/e03d508163710ef8c68513b33496a0ec5e99ef96/src/replicator.rs#L78 and https://github.com/Nao-ris/couchbase-lite-rust/blob/e03d508163710ef8c68513b33496a0ec5e99ef96/src/replicator.rs#L86 should we retain, the ref comes from a C call

File: tests/replicator_tests.rs [hunk 1]
```diff
@@ -76,67 +78,257 @@ fn config() {
...
+        // Save doc
+        utils::add_doc(local_db1, "foo", 1234, "Hello World!");
+
+        // Check the replication process is not pushing to central
+        assert!(!utils::check_callback_with_wait(|| central_db.get_document("foo").is_ok(), None));
+    });
+}
+
+#[test]
+fn push_type_not_pulling() {
```
Comment (MatthieuClochardDoctolib) [inline tests/replicator_tests.rs:120]: Likewise shouldn't we test that local_db2 is still pushing?

**Review by MatthieuClochardDoctolib — CHANGES_REQUESTED** [review]
Active waits in test are kind of an hassle, the runtime is quite long :/ Not sure we can do better to test replication though.

File: src/field_encryption.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    pub fn create_with_double(value: f64) -> Encryptable {
+        unsafe { CBLEncryptable_CreateWithDouble(value).into() }
+    }
+
+    pub fn create_with_string(value: String) -> Encryptable {
+        unsafe {
+            let stri = value.as_str();
+            println!("Antoine building: {:?}", stri);
```
[the remark this answers is not in the document]
Comment (Nao-ris) [reply src/field_encryption.rs:49]: :-)

File: tests/replicator_tests.rs [hunk 3]
```diff
@@ -76,67 +78,257 @@ fn config() {
         assert_eq!(proxy.port, 3000);
         assert_eq!(proxy.username, Some("username".to_string()));
         assert_eq!(proxy.password, Some("password".to_string()));
-        assert_eq!(repl_config_out.headers, headers_mut.as_dict());
+        assert_eq!(repl_config_out.headers, HashMap::new());
```
[the remark this answers is not in the document]
Comment (Nao-ris) [reply tests/replicator_tests.rs:81]: The repl_config_in is consumed when into() is called

File: tests/replicator_tests.rs [hunk 4]
```diff
@@ -76,67 +78,257 @@ fn config() {
...
 
-        // Start replication
-        repl.start(false);
+        // Check if replication to DB 2
+        assert!(utils::check_callback_with_wait(|| local_db2.get_document("foo").is_ok(), None));
+    });
+}
+
+#[test]
+fn pull_type_not_pushing() {
```
[the remark this answers is not in the document]
Comment (Nao-ris) [reply tests/replicator_tests.rs:103]: Yes

**Pull request merged**